hack/generator/pkg/astmodel/resource_type.go

```diff
@@ -286,26 +288,73 @@ func (resource *ResourceType) EmbeddedProperties() []*PropertyDefinition {
 	}
 }
 
+// WithProperty creates a new ResourceType with another property attached to it
+// Properties are unique by name, so this can be used to Add and Replace a property
+func (resource *ResourceType) WithProperty(property *PropertyDefinition) *ResourceType {
+	if property.HasName("Status") || property.HasName("Spec") {
+		panic(fmt.Sprintf("May not modify property %q on a resource", property.PropertyName()))
+	}
+
+	// Create a copy to preserve immutability
+	result := resource.copy()
+	result.properties[property.propertyName] = property
+
+	return result
+}
+
+// WithoutProperty creates a new ResourceType that's a copy without the specified property
+func (resource *ResourceType) WithoutProperty(name PropertyName) *ResourceType {
+	if name == "Status" || name == "Spec" {
+		panic(fmt.Sprintf("May not remove property %q from a resource", name))
+	}
+
+	// Create a copy to preserve immutability
+	result := resource.copy()
+	delete(result.properties, name)
+
+	return result
+}
+
 // Properties returns all the properties from this resource type
 // An ordered slice is returned to preserve immutability and provide determinism
 func (resource *ResourceType) Properties() []*PropertyDefinition {
 
-	specProperty := NewPropertyDefinition("Spec", "spec", resource.spec).
-		WithTag("json", "omitempty")
-
 	result := []*PropertyDefinition{
-		specProperty,
+		resource.createSpecProperty(),
 	}
 
 	if resource.status != nil {
-		statusProperty := NewPropertyDefinition("Status", "status", resource.status).
-			WithTag("json", "omitempty")
-		result = append(result, statusProperty)
+		result = append(result, resource.createStatusProperty())
 	}
 
 	return result
```

Comment:
Shouldn't `Properties()` return the non-spec/status properties too?

Reply:
And thus we prove that I need to write the test even for code I think I can't possibly get wrong! 😊

Reply:
Fixed.